src/mame/sinclair/pentevo.cpp

```diff
@@ -0,0 +1,693 @@
+			u16 pap = (*attr >> 3) & 0x0f;
+			u8 pix8 = (invert_attrs && (*attr & 0x80)) ? ~*scr : *scr;
+
+			for (u8 b = (0x80 >> (x % 8)); b; b >>= 1, x++, hpos++)
+				*pix++ = (pix8 & b) ? ink : pap;
+			scr++;
+			attr++;
+		}
+	}
+}
+
+void pentevo_state::pentevo_update_screen_zx16(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+	// TODO attrs not decoded
+	for (u16 vpos = cliprect.top(); vpos <= cliprect.bottom(); vpos++)
+	{
+		u16 hpos = cliprect.left();
+		u16 x = hpos - get_screen_area().left();
+		u16 y = vpos - get_screen_area().top();
+		u8 *scr = &m_screen_location[((y & 7) << 8) | ((y & 0x38) << 2) | ((y & 0xc0) << 5) | (x >> 3)];
+		u16 *pix = &(bitmap.pix(vpos, hpos));
+
+		while (hpos <= cliprect.right())
+		{
+			u16 ink = 0;
+			u16 pap = 7;
+			u8 pix8 = *scr;
+
+			for (u8 b = (0x80 >> (x % 8)); b; b >>= 1, x++, hpos++)
+				*pix++ = (pix8 & b) ? ink : pap;
+			scr++;
+		}
+	}
+}
+
+void pentevo_state::pentevo_update_screen_tx(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+	u8* screen_location = m_ram->pointer() + ((BIT(m_port_7ffd_data, 3) ? 10 : 8) << 14);
+	for (u16 vpos = cliprect.top(); vpos <= cliprect.bottom(); vpos++)
+	{
+		u16 y = vpos - get_screen_area().top();
+		for (u16 hpos = cliprect.left() & 0xfff8; hpos <= cliprect.right();)
+		{
+			u16 x = hpos - get_screen_area().left();
+			u8 *symb_location = screen_location + 0x1c0 + (x >> 4) + ((y >> 3) * 64);
+			u8 *attr_location = symb_location + 0x2000 + BIT(x, 3);
+			if (BIT(x, 3))
+				symb_location += 0x1000;
+			else
+				attr_location += 0x1000;
+
+			u8 attr = *attr_location;
+			u8 fg = ((attr & 0x40) >> 3) | (attr & 0x07);
+			u8 bg = (((attr & 0x80) >> 1) | (attr & 0x38)) >> 3;
+
+			u8 chunk = m_char_ram->read((*symb_location << 3) + (y & 0x07));
+			for (u8 i = 0x80; i; i >>= 1)
+			{
+				bitmap.pix(vpos, hpos++) = (chunk & i) ? fg : bg;
+			}
+		}
+	}
+}
+
+u8 pentevo_state::nemo_ata_r(u8 cmd)
+{
+	bool data_read = (cmd & 0x7) == 0;
+	u8 data;
+	if (data_read && m_ata_data_hi_ready)
+	{
+		data = m_ata_data_latch;
+		m_ata_data_hi_ready = false;
+	}
+	else
+	{
+		data = atm_state::ata_r(cmd << 5);
+		m_ata_data_hi_ready = data_read;
+	}
+
+	return data;
+}
+
+void pentevo_state::nemo_ata_w(u8 cmd, u8 data)
+{
+	bool data_write = (cmd & 0x7) == 0;
+	if (data_write && !m_ata_data_hi_ready)
+	{
+		m_ata_data_latch = data;
+		m_ata_data_hi_ready = true;
+	}
+	else
+	{
+		atm_state::ata_w(cmd << 5, data);
+		m_ata_data_hi_ready = false;
+	}
+}
+
+u8 pentevo_state::spi_port_77_r(offs_t offset)
+{
+	return is_shadow_active() ? 0xff : 0x00;
+}
+
+void pentevo_state::spi_port_77_w(offs_t offset, u8 data)
+{
+	if (is_shadow_active())
+		return atm_port_ff77_w(0x77 | offset, data);
+
+	m_sdcard->spi_ss_w(BIT(data, 0));
+	m_zctl_cs = BIT(data, 1);
+}
+
+u8 pentevo_state::spi_port_57_r(offs_t offset)
+{
+	if (m_zctl_cs)
+		return 0xff;
+
+	u8 din = m_zctl_di;
+	spi_port_57_w(0, 0xff);
+	return din;
```

Comment:
This also needs a check for `machine().side_effects_disabled()`.

Reply:
added in spi_port_57_w()